app/src/main/java/com/alphawallet/app/viewmodel/BaseViewModel.java

```diff
@@ -1,109 +1,176 @@
-			}
-			else
-			{
-				error.postValue(new ErrorEnvelope(C.ErrorCode.UNKNOWN, message, throwable));
-			}
-		}
-	}
-
-	public static void onQueueUpdate(int complete)
-	{
-		queueCompletion.postValue(complete);
-	}
-
-	public static void onPushToast(String message)
-	{
-		pushToastMutable.postValue(message);
-	}
-
-	public void showSendToken(Context context, String address, String symbol, int decimals, Token token) {
-		//do nothing
-	}
-
-	public void showTokenList(Activity activity, Token token) {
-		//do nothing
-	}
-
-	public void showErc20TokenDetail(Activity context, String address, String symbol, int decimals, Token token) {
-		//do nothing
-	}
+    protected static final MutableLiveData<Integer> queueCompletion = new MutableLiveData<>();
+    protected static final MutableLiveData<String> pushToastMutable = new MutableLiveData<>();
+    protected static final MutableLiveData<Integer> successDialogMutable = new MutableLiveData<>();
+    protected static final MutableLiveData<Integer> errorDialogMutable = new MutableLiveData<>();
+    protected static final MutableLiveData<Boolean> refreshTokens = new MutableLiveData<>();
+    protected final MutableLiveData<ErrorEnvelope> error = new MutableLiveData<>();
+    protected final MutableLiveData<Boolean> progress = new MutableLiveData<>();
+    protected Disposable disposable;
+    private AnalyticsServiceType analyticsService;
+
+    public static void onQueueUpdate(int complete)
+    {
+        queueCompletion.postValue(complete);
+    }
+
+    public static void onPushToast(String message)
+    {
+        pushToastMutable.postValue(message);
+    }
+
+    @Override
+    protected void onCleared()
+    {
+        cancel();
+    }
+
+    private void cancel()
+    {
+        if (disposable != null && !disposable.isDisposed())
+        {
+            disposable.dispose();
+        }
+    }
+
+    public LiveData<ErrorEnvelope> error()
+    {
+        return error;
+    }
+
+    public LiveData<Boolean> progress()
+    {
+        return progress;
+    }
+
+    public LiveData<Integer> queueProgress()
+    {
+        return queueCompletion;
+    }
+
+    public LiveData<String> pushToast()
+    {
+        return pushToastMutable;
+    }
+
+    public LiveData<Boolean> refreshTokens()
+    {
+        return refreshTokens;
+    }
+
+    protected void onError(Throwable throwable)
+    {
+        Timber.tag("TAG").d(throwable, "Err");
+        if (throwable instanceof ServiceException)
+        {
+            error.postValue(((ServiceException) throwable).error);
+        }
+        else
+        {
+            String message = throwable.getMessage();
+            if (TextUtils.isEmpty(message))
+            {
+                error.postValue(new ErrorEnvelope(C.ErrorCode.UNKNOWN, null, throwable));
+            }
+            else
+            {
+                error.postValue(new ErrorEnvelope(C.ErrorCode.UNKNOWN, message, throwable));
```

Comment:
Move `error.postValue(...` out of `if else`. Get 2 benefits:
1. Eliminate duplication
2. Easier to understand what is `if else` for

Reply:
I think this is very old code, but got highlighted when I formatted it. But awesome catch nonetheless! :)

Reply:
Done!